### ISSUE TITLE
Use feature gate on parallelism

### ISSUE THREAD

**Issue by hyf0** [oxc-project/oxc#2959]
Related:
 - https://github.com/rolldown/rolldown/pull/845

Currernt `oxc_sourcemap` enable using rayon by default and it show improved performance said in https://github.com/oxc-project/oxc/pull/2825#discussion_r1542740699.

However, `wasm` target doesn't have a good support in rayon. While rayon claims it's ok in some issues, but it's still quite frustrating to debug wasm + rayon. It would be helped if we could disable rayon using feature gate.

---

Off the topic,  Using feature gate on parallelism should be a principle that apply on all oxc crates that gonna exposed to users when introducing any parallelism in it.

**Comment by hyf0** [comment]
I also have some confusions on https://github.com/oxc-project/oxc/pull/2825#discussion_r1542740699. While it said it has improved the performance overall, but it's doesn't it's suitable strategy on evert situation.

I don't much about these two funcitons:

https://github.com/oxc-project/oxc/blob/6561392202f8dd739ed3fbdb3ea409b9c0a0a2e0/crates/oxc_sourcemap/src/encode.rs#L11

https://github.com/oxc-project/oxc/blob/6561392202f8dd739ed3fbdb3ea409b9c0a0a2e0/crates/oxc_sourcemap/src/encode.rs#L74

But it seems that gonna be used in codegen. Then I realized these are gochas in logic

- Rolldown is rendering modules, which means using oxc_codegen, in parallel already, so if this parallelism is enable by default , I'm not sure if it really improve performance while we are already apply parallelism in module level.

- Rolldown will use `oxc_sourcemap` directly to manipulate the sourcemap in chunk level. Which means apply these functions on modules in chunk in serial.

I suspect that where it really improve the performance since the benchmark case used in https://github.com/oxc-project/oxc/pull/2825#discussion_r1542740699, which is a huge single chunk.

If we have a case that generate so many mutiple small chunks, I wonder if it would be even a downside to the performance.

However, it still show pentation on the case said https://github.com/oxc-project/oxc/pull/2825#discussion_r1542740699. So using feature gate might not enough, since we need dynamicly decide to use which method depend on the input.

**Comment by Boshen** [comment]
Let's either:

* remove rayon from this case
* add benchmark to indicate real world scenarios

**Comment by hyf0** [comment]
Let me try to find a way to benchmark this. On hold for now.

**Comment by Boshen** [comment]
Next release:

https://github.com/oxc-project/oxc/blob/f6daf0b7ae4fb9a3ac9e8520824eb06a278c78ea/crates/oxc/Cargo.toml#L43-L44
